src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/ConsoleDebugTargetsProvider.cs

```diff
@@ -321,24 +316,103 @@ public string GetFullPathOfExeFromEnvironmentPath(string exeToSearchFor)
+        /// <param name="unescaped">The string to escape.</param>
+        /// <param name="toEscape">The characters to escape in the string.</param>
+        /// <returns>The escaped string.</returns>
+        internal static string EscapeString(string unescaped, char[] toEscape)
+        {
+            if (string.IsNullOrWhiteSpace(unescaped)) return unescaped;
+
+            var currentState = StringState.NoSlash;
+            var finalBuilder = new StringBuilder();
+            foreach (var currentLetter in unescaped)
+            {
+                switch (currentState)
+                {
+                    case StringState.NoSlash:
+                        // If we're currently not in a quoted string, then we need to escape anything in toEscape.
+                        // The valid transitions are to OneSlash and InString. Any '"' encountered here is a raw quote.
+                        if (currentLetter == '\\')
+                        {
+                            currentState = StringState.OneSlash;
+                        }
+                        else if (currentLetter == '"')
+                        {
+                            currentState = StringState.InString;
+                        }
+                        else if (toEscape.Contains(currentLetter))
+                        {
+                            finalBuilder.Append('^');
+                        }
+
+                        finalBuilder.Append(currentLetter);
+                        break;
+                    case StringState.OneSlash:
+                        // If a '\' was the previous character, then we blindly append to the string, escaping if necessary,
+                        // and move back to NoSlash.
+                        if (toEscape.Contains(currentLetter))
+                        {
+                            finalBuilder.Append('^');
+                        }
+
+                        finalBuilder.Append(currentLetter);
+                        currentState = StringState.NoSlash;
+                        break;
+                    case StringState.InString:
+                        // If we're in a string, we don't escape any characters. If the current character is a '\',
+                        // then we move to InStringOneSlash. Otherwise, we're still in the string.
+                        if (currentLetter == '\\')
+                        {
+                            currentState = StringState.InStringOneSlash;
+                        }
+                        else if (currentLetter == '"')
+                        {
+                            currentState = StringState.NoSlash;
+                        }
+
+                        finalBuilder.Append(currentLetter);
+                        break;
+                    case StringState.InStringOneSlash:
+                        // If we have one slash, then we blindly append to the string, no escaping, and move back to
+                        // InString.
```

Comment:
I find it helps to give examples, for example, took a little to understand that this is handling when you escape a quote within a quoted string.